blogs/2023/07/20/mangling-vscode.md

```diff
@@ -0,0 +1,177 @@
+
+That calculus changes if we can get nice optimizations like this essentially for free, say by having our build tool do them for us automatically. And indeed, smart tools like esbuild already implement identifier mangling. That means we can keep writing our `veryLongAndDescriptiveNamesThatWouldMakeEvenObjectiveCProgrammersBlush` and let our build tools shorten them for us!
+
+Even though esbuild implements mangling, by default it only mangles names when it is reasonably confident that mangling won't change the behavior of the code. After all, having a bundler break your code really stinks. In practice, this means that esbuild mangles local variable names and argument names. This is safe unless your code is doing some truly crazy stuff (in which case, you likely have far bigger issues than code size to worry about).
+
+However esbuild's conservative approach means that it skips mangling many names because it can't be confident that changing them is safe. As a simple example of how things could go wrong, consider:
+
+```ts
+const obj = { longPropertyName: 123 };
+
+function lookup(prop) { return obj[prop]; }
+
+console.log(lookup('longPropertyName'));
+```
+
+If mangling changes `longPropertyName` to `x`, the dynamic lookup on the next line will no longer work:
+
+```ts
+const obj = { x: 123 }; // Here `longPropertyName` gets rewritten to `x`
+
+function lookup(prop) { return obj[prop]; }
+
+console.log(lookup('longPropertyName')); // But this reference doesn't and now the lookup is broken
+```
+
+Notice in the code above how we're still trying to use `longPropertyName` to access the property even though the property itself has been changed during mangling.
+
+While this example is contrived, there are actually many ways these breaks can happen in real code:
+
+- Dynamic property access.
+- Serializing objects or parsing JSON to an expected object shape.
+- APIs you expose (the consumers won't know about the new mangled names.)
+- APIs you consume (including DOM APIs.)
+
+Although you can force esbuild to mangle basically every single name it finds, doing so completely breaks VS Code for the reasons outlined above.
+
+Despite this, I couldn't shake the feeling that we must be able to do better in the VS Code codebase.  If we couldn't mangle every name, perhaps we could at least find some subset names we safely could.
+
+## False starts with private props
+
+Looking back over our minified sources, another thing that jumped out at me was how many long names I saw starting with `_`. By convention, this indicates a private property. Surely private properties can be safely mangled and code outside of the class would be none the wiser, right? And wait, shouldn't esbuild be doing this already for us? Yet I knew that the folks who wrote esbuild are no slouches. If esbuild wasn't mangling private properties, it was almost certainly for good reason.
```

Comment:
Maybe mention that this is another esbuild mode that is off by default and at your own risk

Reply:
I actually call this out later on. It was how I first prototyped mangling private props